app/controllers/devise/two_factor_authentication_controller.rb

```diff
@@ -63,43 +77,44 @@ def handle_valid_otp
       bypass_sign_in resource
       flash[:notice] = t('devise.two_factor_authentication.success')
 
-      update_metrics
+      analytics.track_event('User 2FA successful')
 
       resource.update(second_factor_attempts_count: 0)
 
       redirect_valid_resource
     end
 
-    def update_metrics
-      analytics.track_event('User 2FA successful')
-    end
-
     def redirect_valid_resource
       redirect_to after_sign_in_path_for(resource)
     end
 
-    def show_direct_otp_prompt
-      @code_value = current_user.direct_otp if FeatureManagement.prefill_otp_codes?
+    def prompt_for_otp_reentry
+      if current_otp_delivery_method == :totp
+        show_totp_prompt
+      else
+        show_direct_otp_prompt
+      end
+    end
```

Comment:
For readability, could you please move this method below `handle_second_factor_locked_resource`, since that's the order in which it gets called? It's a best practice to write code in a top-down narrative. When I'm reading code, and a method calls another method, I feel it's more natural to look for that method below the one that called it. 

Here's an excerpt from the book "Clean Code": https://books.google.com/books?id=hjEFCAAAQBAJ&pg=PA37&lpg=PA37&dq=clean+code+top-down+narrative&source=bl&ots=T6iIxIYIKp&sig=TuzVgMANvt8Uv6qqkeWrkrXsGoY&hl=en&sa=X&ved=0ahUKEwia26PvsOzOAhXKPB4KHZcSBUYQ6AEIIDAB#v=onepage&q=clean%20code%20top-down%20narrative&f=false